### DISCUSSION TITLE
Communicaton between Veramo agent and credential exchange

### DISCUSSION THREAD

**Discussion by andra961** [decentralized-identity/veramo#733] [Documentation]
Hi everyone, i noticed that(if i'm not mistaken) in the documentation there aren't guides when it comes to instaurate communication between agents. My goal is to develop a prototype of credential exchange between an issuer and a holder(probably with veramo react native) by using for example qr codes. Would it be possible with the current state of the framework? and from where should i start implementing it?any help would be greatly appreciated, thx in advance.

**Comment by mirceanis** [comment]
You are right, we don't have an example of end to end communication between agents. This is mostly because there is no one size fits all solution that we could present as the default.

**Option 1: DIDComm**
Since Veramo deals with DIDs, the natural communication channel between DIDs is through DIDComm (`@veramo/did-comm`), however there are caveats.
The DID documents of the agents involved (at the very least, the receiver DID) need to contain 2 things. An X25519 `keyAgreement` key and a `DIDCommMessaging` endpoint.
Also, the receiving agent needs to listen to messages sent to that endpoint to get the messages.
To get an idea of how this looks like in code, there is a [sample in the integration test suite](https://github.com/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/__tests__/shared/didcommRemote.ts#L33) where an agent sends a message to itself between 2 DIDs controlled locally.

If you use `did:web`, this comes [mostly out of the box in veramo](https://github.com/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/packages/remote-server/src/default-did.ts#L40). You still need to configure the hostname, though.

**Option 2+: A custom protocol**
To establish a communication channel, you would likely have the sender find out how and where to send something.
For that to work you could, for example, have the receiver generate some channel information and present that as a QR code.
To get started, it can be as simple as a URL where the sender can send HTTP POST messages.
This can already be done using [`@veramo/remote-server`](https://github.com/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/packages/remote-server/src/messaging-router.ts#L32)
This creates an expressjs HTTP POST router that pipes the message into veramo's `handleMessage()` logic.

Other options are possible too, but they require a lot more work than can be detailed here.
- use WalletConnect
- use libp2p
- use some centralized service like firebase
- ...

In any case, you need a few things:
- receiver advertising where it's listening (and maybe also an encryption key). In the case of `@veramo/remote-server`, for example, that's just knowing the URL of the server.
- a messageHandler plugin used with the `@veramo/message-handler` to make sense of the incoming package. If your messages are plain credentials or presentations, then you don't need to do anything, actually. Veramo already supports receiving direct VC/VP payloads.
- some logic on the sender side to encode/encrypt the stuff to be sent.

I'm stopping now before going too deep in the rabbit hole. Perhaps you can share more details about your constraints I can guide you to more relevant answers.
I hope this helps

**Reply by andra961** [reply]
Thanks, your answer certainly does help. For now i'd restrict it to `DIDcomm.` As i said, my purpose for agent communication would be that of issuing credentials and requesting presentations of said credentials. Consider this example: a user downloads an app made with veramo-react-native and uses it to receive or authenticate through credentials on services built with veramo. I also have seen the code example you linked and i'd have a few questions. 

1) Assuming i am operating on an ethereum net and thus saving/modifyng the `ethr-did-registry` smart-contract on it: What steps do i actually need to setup a communication like the one of the tests?do i just have to call the `addSevicePoint` function with an endpoint(ex: localhost) and add it on the agents' did documents? or do i have to manually set some server?

2) does Veramo workflow in the matter of exchanging credentials assume they are always saved locally in wallets?or are/can/should they  be referenced in the blockchain somehow(maybe through the did-registry attributes?)

3)some doubts on the `didManagerImport` function:
how would i permanently allow a user to gain access to a specific did and making him able to reimport it each time he downloads a veramo app?

Sorry for all the doubts, and thanks again for your time.

**Reply by mirceanis** [reply]
These are all excellent questions, and they help us guide the codebase to a more usable state.
I will try to answer each one:

1. It depends.
If you are just running local tests on your computer, it is fine to use localhost as a messaging endpoint. This will not work accross devices, though, so the next easy option is to use something like https://ngrok.com/ which gives you public URLs to servers running on your local machine.
If you want to use `did:ethr:rinkeby` as the identifiers for your users then, when you call [`agent.didManagerAddService()`](https://github.com/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/__tests__/shared/didManager.ts#L179) to advertise a service endpoint, a transaction will be sent to the rinkeby testnet. The account has to have some funds for the transaction costs. You can use any other ethereum network, but a testnet helps you avoid costs during testing.
If all of that is too complicated, it may be easier to start with `did:web`. In this case a webserver is needed with https capability, and this is where **ngrok** can help a lot!

2. Credentials don't necessarily have to be saved in wallets, but we would strongly discourage saving them to a blockchain.
Our examples use a local database for credentials and all the other storage requirements of an agent, but that can technically be replaced with other storage options.
There is even the option to use remote storage (like a centralized service), either as a plain remote database in a cloud, or as a remote veramo agent that only exposes storage methods to other agents ( this can be done with `@veramo/remote-client` & `@veramo/remote-server`)

3. To ensure that each user can recover their DID and keys when they change devices or reinstall, the easiest option is to have a deterministic way of generating keys. Our current examples don't show this, since keys are generated randomly by `@veramo/kms-local` but you can technically generate keys externally using a mnemonic phrase and import them. We are planning to add a mnemonic based key management system example, but it's not really high on the priority list yet.

In all of the above examples, the user agent is expected to be able to host a service endpoint, which is probably not easy to do in a react-native environment. To work around this issue, the service endpoint that user DIDs advertise can be a centralized solution that can forward messages to the mobile app. Then the app calls `agent.handleMessage()` with the payload and veramo takes care of the rest.
Alternatively, you could set up an agent in some cloud that acts as an inbox service for your mobile agents and your mobile app later pulls the messages from the inbox when it starts, for example.

We would be very interested in getting more external contributions to this project; so, if you have ideas about simplifying all of the above please share.

**Reply by andra961** [reply]
Thank you very much for your kind answer. For now i'm still trying to get the gist of it, and if you don't mind i'd still have some doubts. 
1)In the integration test you provided it's not very clear to me when the association between the endpoint and the agent itself is made. How does the agent know to which endpoint listen for incoming messagges and how is the bound made(apart from adding the endpoint in the DID document)?

2)I was also wondering, since DIDcomm appears to work just by knowing the DID of the other agent, is `agent.resolveDID` wrapped in `agent.sendDIDCommMessage` and called to get the endpoint of the receiver given the DID?

3)Assuming i would like to test a VC exchange between two agents in local using something like Ganeche CLI, how could i use one of the provided and founded addresses to add the servicePoints, given that Veramo generates its DIDs randomly?DID which, from what i understand need to be founded in order to add a service enpoint to them and use them in a DIDcomm exchange. 

Once again thank you for your patient, and i hope this helps you pinpoint the darkest points of the Veramo workflow, given the current documentation.

**Reply by mirceanis** [reply]
Thank you for being so thorough with your questions!
I will try to provide more clarity for each one.

1) In the case of `did:web`, the agent can use [web-did-doc-router](https://github.dev/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/packages/remote-server/src/web-did-doc-router.ts#L30) to serve the DID document for a particular DID.
And, for messaging, you can use the [messaging-router](https://github.dev/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/packages/remote-server/src/messaging-router.ts#L32) to create an endpoint that can listen for messages.
These are expressJS routers, so they depend on the expressJS app setup.

For example, in the [restAgent.test](https://github.dev/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/__tests__/restAgent.test.ts#L198) There is a `MessagingRouter` added that listens to `<basePath>/messaging` and forwards whatever is posted there to the `agent.handleMessage()` function.

This agent is then listening at that endpoint URL regardless of the fact that the URL is posted to a DID document or not.
To let the world know that there is an agent listening for messages there, this endpoint can be advertised using `agent.didManagerAddService()`. This is where the association is made to the DID.

Note that there is no one to one mapping between endpoints and DIDs or DID document service entries.
The same endpoint can technically be used by multiple DIDs that are managed by the agent, but for simplicity it's probably wiser to separate them somehow.

2) Exactly. The receiver DID Document is automatically resolved when you send a message.

3) One option is to [initialize ganache with a known set of private keys](https://github.dev/decentralized-identity/ethr-did-resolver/blob/9693c9048d6940c3eff77638419d4e0dd95f2a82/src/__tests__/testUtils.ts#L6) and then import the [same keys during your agent setup](https://github.com/decentralized-identity/ethr-did-resolver/blob/9693c9048d6940c3eff77638419d4e0dd95f2a82/src/__tests__/resolver.test.ts#L38).
Another option is to list the addresses that your agent is using for did:ethr and then funding these addresses from the pre-funded accounts in ganache.

My answers are perhaps a little vague and the solutions I provide may be overly complicated because of the nature of this framework. We want it to be un-opinionated and allow you to use whatever setup your application or system needs.
The downside of this approach is that there are no rails to guide you.

One possible plan we have to address this is to create several opinionated setups; recipes for particular situations. Since we don't have the bandwidth to explore all sorts of situations we rely on members of the community to surface the more common use-cases.
After you manage to put all the pieces in place for your demo, consider sharing your setup with the community here, as surely someone else will run into these questions.

**Reply by andra961** [reply]
My plan is to use did:ethr, not did:web. Since i asked you this questions, i managed to import through private keys some funded eth addresses in veramo and succesfully add localhost endpoints to them through `agent.didManagerAddService()`. However, i get an error(`Error: key_not_found: could not map an agent key to an skid for did:ethr`) when i try to send a message between them through DIDcomm, and i noticed it's probably because of the lack of the `keyAgreement` section in the did documents used. How could i add key agreement to my dids?is it through `agent.didManagerAddKey()` or should i use an other method?

**Reply by mirceanis** [reply]
That's great news.
On the sender side, you can avoid using key agreement keys by packing your messages with `anoncrypt` instead of `authcrypt`.
Or, just to get your endpoints setup you can use plaintext messages at first.

The veramo DIDComm implementation can automatically transform Ed25519 signing keys into X25519 encryption keys (for keyAgreement).
This sample [here](https://github.dev/uport-project/veramo/blob/828bfa20c6181c2e3a405ab13a06249f5112eb72/__tests__/shared/dbInitOptions.ts#L115) uses Ed25519 keys in the receiving DID document.
The same flow should work with `did:ethr`.
If it does not, then we have bugs to fix :)

Edit: I fixed the wrong link in my answer

**Reply by andra961** [reply]
Unfortunately, it's not working. I added `Ed25519` keys on both receiver and sender with `agent.keyManagerCreate` as in your example, followed by `agent.didManagerAddKey`. When i resolve the documents i can see them in `verificationMethod` like this:
{
        id: 'did:ethr.........',
        type: 'Ed25519VerificationKey2018',
        controller: 'did:ethr.........',
        publicKeyHex: 'publickey....'
      }
,
but still i get the message `Error: key_not_found: could not map an agent key to an skid for did:ethr`.

I also tried different packings:

1)`packing: "none"`: i get the following error => ```
Error: Cannot send DIDComm message through transport with id: 'id': Error: Error when sending DIDComm message through transport with 
id: 'id': failed to send message: FetchError: request to http://localhost:3002/messaging failed, reason: connect ECONNREFUSED 127.0.0.1:3002

2)`packing: "anoncrypt"`: i get the following error => 
TypeError: Cannot read property 'alg' of undefined
    at createJWE (project_path\node_modules\@veramo\did-comm\node_modules\did-jwt\src\JWE.ts:98:21)

Hope this helps

**Reply by mirceanis** [reply]
Can you try to post a receiver DID document as you resolve it?

I think error 1) is indicating a problem with the setup.
error 2) is probably related to the `key_not_found` message and is likely a problem with Veramo.

In all cases, it's much easier to work with code than to imagine how your setup looks like; so it's helpful if you have a project you can link to.

**Reply by andra961** [reply]
The didDocument of the receiver looks like this:

```
{
  didDocumentMetadata: { versionId: '13103712', updated: '2021-11-08T09:26:03Z' },
  didResolutionMetadata: { contentType: 'application/did+ld+json' },
  didDocument: {
    '@context': [
      'https://www.w3.org/ns/did/v1',
      'https://identity.foundation/EcdsaSecp256k1RecoverySignature2020/lds-ecdsa-secp256k1-recovery2020-0.0.jsonld'
    ],
    id: 'did:ethr:...',
    verificationMethod: [
      {
        id: 'did:ethr:...#controller',
        type: 'EcdsaSecp256k1RecoveryMethod2020',
        controller: 'did:ethr:...',
        blockchainAccountId: 'address of the current owner, since i changed owner once'
      },
      {
        id: 'did:ethr:#delegate-1',
        type: 'gianni(this was just for testing addKey()',
        controller: 'did:ethr:..."
        publicKeyHex: 'publicKeyHex'
      },
      {
        id: 'did:ethr:...#delegate-6',
        type: 'Ed25519VerificationKey2018',
        controller: 'did:ethr:...',
        publicKeyHex: 'public key hex'
      }
    ],
    authentication: [
      'did:ethr:...#controller'
    ],
    assertionMethod: [
      'did:ethr:...#controller',
      'did:ethr:...#delegate-1',
      'did:ethr:...#delegate-6'
    ],
    service: [
      {
        id: 'did:ethr:...#service-1',
        type: 'pippo(again, just testing addService())',
        serviceEndpoint: 'Giacomo\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
      },
      {
        id: 'did:ethr:...#service-3',
        type: 'DIDCommMessaging',
        serviceEndpoint: 'http://localhost:3002/messaging'
      }
    ]
  }
}
```

For the project itself as for now is really just the node tutorial of the documentation to which i imported 2 eth did with private keys through `didManagerImport()`, added endpoints to them and keys as you can see in the didDocument, and tried to use didComm out of the box.

Also, as you can see, the resolved document is omitting some services(service-2) and keys i added to it previously, i wonder why.

**Reply by mirceanis** [reply]
I think I see the problem now (or at least one problem).
The `@veramo/did-provider-ethr` is not adding any keys to the `keyAgreement` section [here](https://github.com/uport-project/veramo/blob/6ab85166fe94f2c7401270ffcd54db35b900c7b6/packages/did-provider-ethr/src/ethr-did-provider.ts#L103).
I think this can be fixed, but in the meantime please try to run your tests with `packing: 'none'` to see if your endpoints are added properly for the agents to communicate in cleartext.

**Comment by andra961** [comment]
Unfortunately, it seems they are not: 

Error: Cannot send DIDComm message through transport with id: '906acdd8-5db2-4a59-a1f6-e81e6bd4a33a': Error: Error when sending DIDComm message through transport with 
id: '906acdd8-5db2-4a59-a1f6-e81e6bd4a33a': failed to send message: FetchError: request to http://localhost:3002/messaging failed, reason: connect ECONNREFUSED 127.0.0.1:3002

Here's the didComm code i'm using:

```
const message = {
        type: 'test',
        to: receiver.did,
        from: sender.did,
        id: 'test',
        body: { hello: 'world' },
      }

      const packedMessage = await agent.packDIDCommMessage({
        packing: "none",
        message,
      })
      const result = await agent.sendDIDCommMessage({
        messageId: '123',
        packedMessage,
        recipientDidUrl: receiver.did,
      })
```

**Reply by mirceanis** [reply]
That error indicates there's no listener on that port (something that a `@veramo/remote-server` + an express app can do). It would be helpful if you could share your configuration.
It's easier to debug if you link to a project.

**Reply by andra961** [reply]
Yeah that's the case, i did not setup any listener since i thought didComm could work out of the box. As i said my configuration right now is very simple and similar to the documentation node setup( but with DIDComm plugin added). Do you have any hint on how to setup the minimal listener to make two ethr agent message on localhost?It would be very appreciated. Also, do you have any news on the errors regarding packing different from "none"?

**Reply by mirceanis** [reply]
I started to write a guide for getting didcomm working with veramo on nodejs [here](https://github.com/uport-project/veramo-website/pull/98)
It's only a draft now, but the first part should have what you need to setup a listener:

```ts
import { MessagingRouter, RequestWithAgentRouter } from '@veramo/remote-server'
import { agent } from './veramo/setup'
import express from 'express'

const port = process.env.PORT || 3002

const messagingRouter = MessagingRouter({
  metaData: { type: 'express' },
})

const app = express()
app
  .use(RequestWithAgentRouter({ agent }))
  .use(messagingRouter)
  .listen(port, async () => {
    console.log(`Local agent listening for messages at ${port}`)
  })
```

Again, it is much easier to help you if you link to a project that contains your setup.
I realize there is a lot missing from our docs and samples, so please help us optimize our time to fix them. The best way to do that is to point to a project with your setup, that is easy to clone and run.
This way you also enable other members of this community to help, in case they have run into similar issues.

**Reply by andra961** [reply]
Thank you very much, this works! i can now send msg with `packing: "none"` locally.
Even if this is already great progess, i would still have some doubts left:

1)Why do you think all the other packing types fail?

2) If setting up an express server is mandatory for receiving incoming messages (and thus credentials) with did:ethr, how would you suggest to approach issuing credentials to a react native veramo agent (mobile wallet)?

3)I also notice that when i add a service endpoint to a did through veramo, after some time it disappears from the document and it's not present anymore when i resolve it. I then have to add an other endpoint in order to use DIDComm (it remains stored in the local db though)

**Reply by mirceanis** [reply]
That's great news!
1) `packing: "jws"` should also work. It is also cleartext, but the message is signed by the sender. The problem with the other packing types is with adding the key to the DID document. Because of a bug, it is only added as a `verificationMethod` and not as a `keyAgreement` key (I don't have an estimate for when it will be fixed as there are other more pressing issues. If you wish to attempt a fix in the meantime, your contributions would be welcome).
After that is fixed, you should be able to use `packing: "anoncrypt"` when only the receiver has a `keyAgreement` key and `packing: "authcrypt"` when both sender and receiver have them.

2) I mentioned this in my first reply and proposed a potential solution:
> In all of the above examples, the user agent is expected to be able to host a service endpoint, which is probably not easy to do in a react-native environment. To work around this issue, the service endpoint that user DIDs advertise can be a centralized solution that can forward messages to the mobile app. Then the app calls agent.handleMessage() with the payload and veramo takes care of the rest.
Alternatively, you could set up an agent in some cloud that acts as an inbox service for your mobile agents and your mobile app later pulls the messages from the inbox when it starts, for example.

There is also a team that is working on a libp2p transport for DIDComm, which might work well on mobile, but there is no estimate for the work.

3) By default, attribute changes for `did:ethr` [expire after 86400](https://github.com/uport-project/ethr-did/blob/afeb431a57d4a2f7d9c641b75dc979897d8b7381/src/index.ts#L157) seconds (1 day). This rather short default was originally chosen to limit the scope of long lived delegation. The idea being that when you delegate signing capabilities to another party, there should be a time window after which it automatically expires.
Users who deliberately need longer delegation should specify a different expiry schedule.
The same default ended up being used for service endpoints as well, where it is rather short and should be changed ([PR welcome](https://github.com/uport-project/veramo/blob/71606ed39be6864573610071941a72e00a3c5612/packages/did-provider-ethr/src/ethr-did-provider.ts#L124)).
You can set a longer expiry for all operations by setting the `ttl` parameter for the `EthrDidProvider` in your agent setup:
```ts
      providers: {
        'did:ethr:rinkeby': new EthrDIDProvider({
          defaultKms: 'local',
          network: 'rinkeby',
          rpcUrl: 'https://rinkeby.infura.io/v3/' + INFURA_PROJECT_ID,
          ttl: 3600 * 24 * 365 * 10, // number of seconds of validity for attribute changes (10 years in this example)
        }),
      }
```

The internal Veramo database does not automatically keep track of the DID attributes as they would be resolved publicly, as it would be quite resource intensive. When there is a need to synchronize the internal DID state with the resolved state, this is done on-demand. This happens, for example, with DIDComm when `skid` needs to be provided.
For did:ethr it would probably be easy to tell when attributes should be updated in the DB, but there are other DID methods that don't have such a predictable behavior, so Veramo will not attempt this automatic tracking.

It occurs to me that if you are using ganache, this might also be caused by restarting the ganache network. In that case, all on-chain attributes get lost while the veramo database file is likely still there.

I hope this helps.

**Reply by andra961** [reply]
Thanks, things are getting clearer and clearer. Though, right now i'd have one doubt left. I now understand the purpose of veramo-server: providing a way (through express) to expose an agent methods remotely with an API interface, other than listening to incoming messages and forwarding them to the agent. What i don't understand though is the role of remote-client. Could you explain a use case for that?Could it perhaps help in forwarding messagges to a react native agent as i want to do or is its purpose totally different?

**Reply by mirceanis** [reply]
TLDR: it's a proxy for calling Veramo agent methods from a different agent as if they were installed locally.

There are multiple ways of setting up a Veramo agent.
The classical way is to run everything locally, as a library of your JavaScript app.
Veramo provides another alternative, which is to host some agent functionality remotely.
In this second situation, there is a server running a Veramo agent that uses `@veramo/remote-server#AgentRouter`. This makes it possible for applications with the right credentials to call these agent methods remotely.
Here's where `@veramo/remote-client` comes in. It is a Veramo plugin that exposes all those remote methods as if they were run locally. To the application, and to other Veramo plugins, the methods provided by the `@veramo/remote-client` look like they are provided by a local plugin.

Example:
[A remote agent](https://github.com/uport-project/veramo/blob/1be5e04e09112c0823d776fe2d55117d71a7b448/__tests__/restAgent.test.ts#L122) exposing its methods to be called by a [local agent plugin](https://github.com/uport-project/veramo/blob/1be5e04e09112c0823d776fe2d55117d71a7b448/__tests__/restAgent.test.ts#L102)
In that integration test suite, all the methods that are called by the tests are called on the local agent, but the actual functionality is implemented by the remote-agent.

One use-case could be a front-end agent that can run credential verification, but relies on a back-end agent to store keys for credential signing. To the front-end it looks like a single agent, but the key material never actually reaches the front-end, only the signed result.
To achieve something like this, the back-end agent would install `@veramo/key-manager`, and hook it up to `@veramo/remote-server#AgentRouter`.
The front-end agent would install all the other plugins, except for key-manager. Instead, it would install `@veramo/remote-client#AgentRestClient` and point it to the back-end URL, with the list of `keyManager*` methods.

I don't know if this sort of setup would be suitable for forwarding messages. In the example from above, the 2 agents act as one. Meaning that if another front-end agent wanted to make use of the same back-end, it would have access to the same messages as the first one.

Note that this possible setup is in early stages of development. The authorization/access-control model for calling remote methods needs a lot of improvement. First of all to define who can call which methods, and then to define what data should be available for various method calls.

I hope this helps

**Reply by mirceanis** [reply]
BTW, we now have [an integration test-suite](https://github.com/uport-project/veramo/blob/next/__tests__/shared/didCommWithEthrDidFlow.ts) that shows how to set up 2 did:ethr that can send messages using DIDComm over HTTP.

It uses ganache to anchor the did:ethr identifiers and a localhost express endpoint for messaging.
It is a single agent looping back messages to itself, but everything would work the same if the DIDs were hosted by different agents.

We had to fix some bugs to make this work. If you want to replicate that setup you will have to install the veramo dependencies from the `@next` dist tag. Example: `yarn add @veramo/core@next`

**Reply by andra961** [reply]
so you fixed the keyAgreement bug?congratulations! About the remote-client i was wondering. First, i notice that in order to plug the client agent to the server one you need its schema, which you already possess locally in the integration test. But how could you access it remotely in a real scenario? Secondly, i was wondering if the client agent also triggers the events which fire on the server one( example, messagge received with access to the message body too) In that case, i think it would be possible to have a remote express agent built for listening to incoming messagges, and a client one on React Native which also receives them through said events, shows them to the user and then he decides how to react to them( example: store the incoming credential or discard it). What do you think?

**Reply by mirceanis** [reply]
The schema json can be posted by the server agent at a known URL, using [`ApiSchemaRouter`](https://github.com/uport-project/veramo/blob/6884b28a334a87b11d9bdf4646995946bbb48b32/packages/remote-server/src/api-schema-router.ts#L46). There is an example of this in the [package docs](https://veramo.io/docs/api/remote-server).
The client can then use the schema json posted by the server.

In the current version events don't propagate back to the client unfortunately. The client does not hold a connection to the server after a method executes, so the server wouldn't know where to send the events.
With this limitation in mind, perhaps a simpler model is to have the server store all messages, and the client can poll for new messages when it comes online, or periodically.

Also, you are not limited by the capabilities of Veramo. For example, you can create your own websocket connection between your react-native app and the server app. The server app can listen to veramo events (like incoming messages) and post them to the websocket. Then, when your react-native app gets an event on the socket, it can react to it locally.

**Reply by andra961** [reply]
Again, thank you for the exhaustive answes. I'll let you know if i have other doubts.

**Reply by mirceanis** [reply]
You're welcome.
I encourage you to tell us more about your project.
There is a Show and Tell section in the discussions that is perfect for this ;)

**Reply by vijay05155** [reply]
Hi everyone, i want to ask something related to '@veramo/remote-server'  this lib.

i was hosted a server using this code that was mention in veramo.io,

import express from 'express'
import { agent } from './agent'
import { AgentRouter, ApiSchemaRouter, WebDidDocRouter } from '@veramo/remote-server'

const getAgentForRequest = async (req: express.Request) => agent
const exposedMethods = agent.availableMethods()
const basePath = '/agent'
const schemaPath = '/open-api.json'

const agentRouter = AgentRouter({
  getAgentForRequest,
  exposedMethods,
})

const schemaRouter = ApiSchemaRouter({
  basePath,
  getAgentForRequest,
  exposedMethods,
})

const didDocRouter = WebDidDocRouter({
  getAgentForRequest,
})

const app = express()
app.use(basePath, agentRouter)
app.use(schemaPath, schemaRouter)
app.use(didDocRouter)
app.listen(3002),

but when i try to use "http://localhost:3002/open-api.json"  this address in postman i got , "agent not avilable"  in response, can you please help me.
